lib/puppet/resource_api.rb

```diff
@@ -166,6 +182,21 @@ def to_resource_shim(resource)
         raise_missing_params if @missing_params.any?
       end
 
+      # If the custom_insync feature is specified but no insyncable attributes are included
+      # in the definition, add the hidden rsapi_custom_insync_trigger property.
+      # This property exists *only* to allow a resource without properties to still execute an
+      # insync check; there's no point in specifying it in a manifest as it can only have one
+      # value; it cannot be specified in a type definition as it should only exist in this case.
+      if type_definition.feature?('custom_insync') && type_definition.insyncable_attributes.empty?
+        custom_insync_trigger_options = {
+          type: 'Enum[do_not_specify_in_manifest]',
+          desc: 'A hidden property which enables a type with custom insync to perform an insync check without specifying any insyncable properties',
+          default: 'do_not_specify_in_manifest',
+        }
+
+        type_definition.create_attribute_in(self, :rsapi_custom_insync_trigger, :newproperty, Puppet::ResourceApi::Property, custom_insync_trigger_options)
```

Comment:
Creating a hidden property seems a bit of a hack. Could RSAPI instead define a [top-level `validate` for the type](https://github.com/puppetlabs/puppet-specifications/blob/master/language/resource_types.md#validate), and check for missing insync properties there? For example `self[<name>]` will be `nil` if a desired value wasn't specified in the manifest.

Reply:
The hidden property is meant to address types which intentionally do _not_ have any insyncable properties - this requirement was originally discovered during exploration for the Puppet Native Testing resources, which don't _ensure_ any state, just validate it; rather than have authors of test resources (or any resource which only has parameters, read only properties, and/or init properties) specify a property which has a default and is never specified in a manifest, rely on this known and documented behavior for traditionally un-insyncable resources.